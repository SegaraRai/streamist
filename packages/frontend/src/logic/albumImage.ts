import { ALBUM_IMAGE_SENTINEL_NODE_IMAGE_ID } from '$shared/dbConfig';
import { dbLinkedListSort } from '$shared/linkedListSort';
import type { AlbumImageWithImage } from '~/types/image';

export function getSortedAlbumImages<T extends AlbumImageWithImage>(
  albumImages: readonly T[]
): T['image'][] {
  if (albumImages.length === 0) {
    return [];
  }

  return dbLinkedListSort(
    albumImages,
    'imageId',
    'nextImageId',
    ALBUM_IMAGE_SENTINEL_NODE_IMAGE_ID
  ).map((item) => item.image);
}

export function getDefaultAlbumImage<T extends AlbumImageWithImage>(
  albumImages: readonly T[]
): T['image'] | undefined {
  return getSortedAlbumImages(albumImages)[0];
}
