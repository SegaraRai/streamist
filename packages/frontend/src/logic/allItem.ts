import type {
  ResourceAlbum,
  ResourceArtist,
  ResourcePlaylist,
  ResourceTrack,
} from '$/types';
import {
  useAllAlbums,
  useAllArtists,
  useAllPlaylists,
  useAllTracks,
} from './useDB';

export type AllItem =
  | {
      t: 'album';
      l: string;
      i: ResourceAlbum;
    }
  | {
      t: 'artist';
      l: string;
      i: ResourceArtist;
    }
  | {
      t: 'playlist';
      l: string;
      i: ResourcePlaylist;
    }
  | {
      t: 'track';
      l: string;
      i: ResourceTrack;
    };

function _useAllItems() {
  const { value: albums } = useAllAlbums();
  const { value: artists } = useAllArtists();
  const { value: playlists } = useAllPlaylists();
  const { value: tracks } = useAllTracks();
  return computed<AllItem[] | undefined>(() => {
    if (!albums.value || !artists.value || !playlists.value || !tracks.value) {
      return;
    }
    return [
      ...albums.value.map(
        (album): AllItem => ({
          t: 'album',
          l: album.title,
          i: album,
        })
      ),
      ...artists.value.map(
        (item): AllItem => ({
          t: 'artist',
          l: item.name,
          i: item,
        })
      ),
      ...playlists.value.map(
        (item): AllItem => ({
          t: 'playlist',
          l: item.title,
          i: item,
        })
      ),
      ...tracks.value.map(
        (item): AllItem => ({
          t: 'track',
          l: item.title,
          i: item,
        })
      ),
    ];
  });
}

export const useAllItems = createSharedComposable(_useAllItems);
