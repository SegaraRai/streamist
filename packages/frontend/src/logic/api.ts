import aspida from '@aspida/axios';
import axios from 'axios';
import createAPI from '$/api/$api';
import { tokens } from './tokens';

const axiosInstance = axios.create();

const api = createAPI(aspida(axiosInstance));

export default api;

export function activateTokenInterceptor() {
  axiosInstance.interceptors.request.use(
    async (config) => {
      const { apiToken } = await tokens.valueAsync;
      if (!config.headers) {
        config.headers = {};
      }
      config.headers.Authorization = `Bearer ${apiToken}`;
      return config;
    },
    (error) => Promise.reject(error)
  );
}
