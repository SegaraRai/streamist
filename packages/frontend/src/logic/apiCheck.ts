export type APIStatus =
  | 'ok'
  | 'ng_needs_auth'
  | 'ng_origin_down'
  | 'ng_cf_down'
  | 'ng_no_connection';

export async function checkAPIStatus(): Promise<APIStatus> {
  if (!navigator.onLine) {
    return 'ng_no_connection';
  }

  try {
    const pingRes = await fetch(`/api/ping?_=${Date.now()}`, {
      method: 'POST',
      redirect: 'manual',
    });

    // Cloudflare Access
    // we cannot get redirected URL
    if (pingRes.type === 'opaqueredirect') {
      return 'ng_needs_auth';
    }

    const pingResText = await pingRes.text();

    if (
      pingRes.status >= 200 &&
      pingRes.status < 300 &&
      pingResText === 'streamist'
    ) {
      return 'ok';
    }

    // Origin down?
    if (
      pingRes.status === 502 ||
      pingRes.status === 503 ||
      pingRes.status >= 520
    ) {
      return 'ng_origin_down';
    }

    // Browser integrity check and CAPTCHA
    if (
      pingResText.includes('cloudflare') &&
      (pingResText.includes('challenge') || pingResText.includes('captcha'))
    ) {
      return 'ng_needs_auth';
    }

    // unknown error
    return 'ng_no_connection';
  } catch (_error: unknown) {
    // do nothing
  }

  return 'ng_no_connection';
}
