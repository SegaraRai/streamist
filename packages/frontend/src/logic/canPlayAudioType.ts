const canPlayTypeCacheMap = new Map<string, CanPlayTypeResult>();

export function canPlayAudioType(type: string): CanPlayTypeResult {
  let result = canPlayTypeCacheMap.get(type);

  if (result == null) {
    result = new Audio().canPlayType(type);
    canPlayTypeCacheMap.set(type, result);
  }

  return result;
}
