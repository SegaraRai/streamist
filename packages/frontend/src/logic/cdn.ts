const DEV_API_CDN_ENDPOINT = `${import.meta.env.VITE_API_ORIGIN}/dev/cdn`;

export const CDN_ENDPOINT = DEV_API_CDN_ENDPOINT;
