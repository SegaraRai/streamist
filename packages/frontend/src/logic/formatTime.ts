/**
 * 時間（時刻ではない）をフォーマットして文字列に変換する \
 * 時間には0以上の値が与えられるものとし、小数点以下は切り捨てられる
 * @example
 * formatTime(0);           // '0:00'
 * formatTime(10);          // '0:10'
 * formatTime(100);         // '1:40'
 * formatTime(1000);        // '16:40'
 * formatTime(100, 100);    // '1:40'
 * formatTime(100, 1000);   // '01:40'
 * formatTime(-1);          // error
 * formatTime(1, -1);       // error
 * formatTime(2, 1);        // error
 * @param time 時間（秒）、0以上
 * @param max 最大時間、`time`以上、この時間をフォーマットしたものに桁や0埋めが合わせられる
 * @return フォーマットされた時間の文字列
 */
export function formatTime(time: number, max?: number): string {
  time = Math.floor(time);

  if (time < 0) {
    throw new Error('`time` must be greater than or equal to 0');
  }

  let hoursLength = 0;
  let minutesLength = 0;

  if (max !== undefined) {
    max = Math.floor(max);

    if (max < 0) {
      throw new Error('`max` must be greater than or equal to 0');
    }

    if (max < time) {
      throw new Error('`max` must be greater than or equal to `time`');
    }

    const maxHours = Math.floor(max / 60 / 60);
    const maxMinutes = Math.min(Math.floor(max / 60), 59);
    hoursLength = maxHours && Math.floor(maxHours).toString().length;
    minutesLength = maxMinutes && Math.floor(maxMinutes).toString().length;
  }

  const hours = Math.floor(time / 60 / 60);
  const minutes = Math.floor(time / 60) % 60;
  const seconds = time % 60;
  const hoursPrefix =
    hoursLength || hours
      ? hours.toString().padStart(hoursLength, '0') + ':'
      : '';
  return `${hoursPrefix}${minutes
    .toString()
    .padStart(minutesLength, '0')}:${seconds.toString().padStart(2, '0')}`;
}
