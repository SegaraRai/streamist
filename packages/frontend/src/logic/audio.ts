import type { TrackFile } from '$prisma/client';
import { canPlayAudioType } from './canPlayAudioType';
import { getTrackFileURL } from './fileURL';

/**
 * `TrackFileDTO`のスコアを計算する \
 * スコアが高いほどユーザーの設定とブラウザの対応状況に即した`TrackFileDTO`であることを表す \
 * 負数は使用不可の（ブラウザが対応していない）フォーマット
 * @param trackFile `TrackFileDTO`
 * @returns スコア
 */
export function calcTrackFileScore(trackFile: TrackFile): number {
  // 再生不可なら-1
  if (!canPlayAudioType(trackFile.mimeType)) {
    return -1;
  }

  const addition = trackFile.extension === '.weba' ? 1_000_000_000 : 0;

  // TODO: ユーザーの音質設定に応じて適切なスコアを出す
  return 1 / trackFile.fileSize + addition;
}

export async function loadAudio(
  audio: HTMLAudioElement,
  trackFiles: readonly TrackFile[],
  play = false
): Promise<void> {
  // スコア降順でTrackFileDTOとスコアの配列を用意
  let trackFilesWithScore = trackFiles
    .map((item) => ({
      score$$q: calcTrackFileScore(item),
      trackFile$$q: item,
    }))
    .sort((a, b) => b.score$$q - a.score$$q);

  // 利用不可なものを除く
  trackFilesWithScore = trackFilesWithScore.filter(
    (object) => object.score$$q >= 0
  );

  // 利用できるものがない場合は例外を送出
  if (!trackFilesWithScore.length) {
    throw new Error('no available trackFile');
  }

  // 一番いいやつを選択
  const trackFile = trackFilesWithScore[0].trackFile$$q;

  // CDNのCookieを設定
  // await fetchAndSetCDNAccessToken();

  const url = await getTrackFileURL(trackFile.trackId, trackFile.id);

  // Audioにsrcを設定
  audio.src = url;
  audio.currentTime = 0;
  if (play) {
    audio.play();
  } else {
    audio.pause();
  }
}
