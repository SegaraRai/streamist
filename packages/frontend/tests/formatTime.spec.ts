import { describe, expect, it } from 'vitest';
import { formatTime } from '~/logic/formatTime';

describe('formatTime', () => {
  it('should throw error if value is less than 0', () => {
    expect(() => formatTime(-1)).toThrowErrorMatchingSnapshot();
  });

  it('should throw error if max is less than 0', () => {
    expect(() => formatTime(0, -1)).toThrowErrorMatchingSnapshot();
  });

  it('should throw error if max is less than value', () => {
    expect(() => formatTime(1, 0)).toThrowErrorMatchingSnapshot();
  });

  it('should return 0:00 for 0 (max < 600)', () => {
    expect(formatTime(0)).toBe('0:00');
    expect(formatTime(0, 0)).toBe('0:00');
    expect(formatTime(0, 599)).toBe('0:00');
  });

  it('should return 00:00 for 0 (600 <= max < 3600)', () => {
    expect(formatTime(0, 600)).toBe('00:00');
    expect(formatTime(0, 3599)).toBe('00:00');
  });

  it('should return 0:00:00 for 0 (3600 <= max < 36000)', () => {
    expect(formatTime(0, 3600)).toBe('0:00:00');
    expect(formatTime(0, 35999)).toBe('0:00:00');
  });

  it('should return 00:00:00 for 0 (36000 <= max < 360000)', () => {
    expect(formatTime(0, 36000)).toBe('00:00:00');
    expect(formatTime(0, 359999)).toBe('00:00:00');
  });

  it('should return 000:00:00 for 0 (360000 <= max < 3600000)', () => {
    expect(formatTime(0, 360000)).toBe('000:00:00');
    expect(formatTime(0, 3599999)).toBe('000:00:00');
  });

  it('should format time correctly', () => {
    expect(formatTime(40)).toBe('0:40');
    expect(formatTime(60)).toBe('1:00');
    expect(formatTime(80)).toBe('1:20');
    expect(formatTime(700)).toBe('11:40');

    expect(formatTime(40, 500)).toBe('0:40');
    expect(formatTime(60, 500)).toBe('1:00');
    expect(formatTime(80, 500)).toBe('1:20');

    expect(formatTime(40, 900)).toBe('00:40');
    expect(formatTime(60, 900)).toBe('01:00');
    expect(formatTime(80, 900)).toBe('01:20');
    expect(formatTime(700, 900)).toBe('11:40');
  });
});
