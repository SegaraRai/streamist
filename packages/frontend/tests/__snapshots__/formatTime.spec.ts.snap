// Vitest Snapshot v1

exports[`formatTime > should throw error if max is less than 0 1`] = `"\`max\` must be greater than or equal to 0"`;

exports[`formatTime > should throw error if max is less than value 1`] = `"\`max\` must be greater than or equal to \`time\`"`;

exports[`formatTime > should throw error if value is less than 0 1`] = `"\`time\` must be greater than or equal to 0"`;
